distributed/comm/ucx.py

```diff
@@ -169,19 +189,44 @@ async def write(
                 frames = await to_frames(
                     msg, serializers=serializers, on_error=on_error
                 )
-                send_frames = [
-                    each_frame for each_frame in frames if len(each_frame) > 0
-                ]
+                nframes = len(frames)
+                cuda_frames = tuple(
+                    hasattr(f, "__cuda_array_interface__") for f in frames
+                )
+                sizes = tuple(nbytes(f) for f in frames)
+                host_frames = host_array(
+                    sum(
+                        each_size
+                        for is_cuda, each_size in zip(cuda_frames, sizes)
+                        if not is_cuda
+                    )
+                )
+                device_frames = device_array(
+                    sum(
+                        each_size
+                        for is_cuda, each_size in zip(cuda_frames, sizes)
+                        if is_cuda
+                    )
+                )
+
+                # Pack frames
+                host_frames_view = memoryview(host_frames)
+                device_frames_view = as_numba_device_array(device_frames)
+                for each_frame, is_cuda, each_size in zip(frames, cuda_frames, sizes):
+                    if each_size:
+                        if is_cuda:
+                            each_frame_view = as_numba_device_array(each_frame)
+                            device_frames_view[:each_size] = each_frame_view[:]
```

Comment:
Do you know what this translates to in Numba under the hood? We could likely implement this `__setitem__` in RMM on `DeviceBuffer` if desired which typically reduces the overhead quite a bit versus Numba.

Reply:
I don't.

We probably could. Would rather wait until we have run this on some real workloads so we can assess and plan next steps.

Reply:
Yeah after poking at this a bit. I think you are right. It's worth adding `__setitem__` to `DeviceBuffer`.

Reused the MRE from issue ( https://github.com/rapidsai/ucx-py/issues/402 ) as a quick test. The bar selected is for the `__setitem__` calls in `send`. There is another one for `recv`. Should add `as_cuda_array` doesn't appear cheap either. Avoiding these seems highly desirable.

<img width="1792" alt="Screen Shot 2020-04-20 at 9 40 44 PM" src="https://user-images.githubusercontent.com/3019665/79826138-c6409e00-834f-11ea-9c7b-e91cd7ba3ed2.png">

Reply:
Should add we would need `__getitem__` as well.

Reply:
Did a rough draft of this in PR ( https://github.com/rapidsai/rmm/pull/351 ).

Reply:
Changed the logic here a bit so that CuPy will be used if available, but will fallback to Numba if not. Am not seeing the same issue with CuPy.